In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [2]:
df = pd.DataFrame(pd.read_csv('base.csv'))

In [ ]:
df.head(5)

In [ ]:
df.info()

In [3]:
# Para mostrar todas as linhas e colunas do dataset

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

In [4]:
df.isnull().mean().round(4)*100

Id                       0.00
ClasseImovel             0.00
ClasseZona               0.00
Fachada                 17.74
TamanhoLote              0.00
Rua                      0.00
Beco                    93.77
FormaProp                0.00
PlanoProp                0.00
Servicos                 0.00
ConfigLote               0.00
InclinacaoLote           0.00
Bairro                   0.00
Estrada1                 0.00
Estrada2                 0.00
TipoHabitacao            0.00
EstiloHabitacao          0.00
Qualidade                0.00
Condicao                 0.00
AnoConstrucao            0.00
AnoReforma               0.00
TipoTelhado              0.00
MaterialTelhado          0.00
Exterior1                0.00
Exterior2                0.00
TipoAlvenaria            0.55
AreaAlvenaria            0.55
QualidadeCobertura       0.00
CondicaoExterna          0.00
TipoFundacao             0.00
AlturaPorao              2.53
CondicaoPorao            2.53
ParedePorao              2.60
TipoAcabPo

In [5]:
# as variáveis Beco, QualdiadeLareira, QualidadePiscina, QualidadeCerca, Outros serão excluídas devido ao alto indice de 
# valores nulos

df.drop(['Beco', 'QualdiadeLareira', 'QualidadePiscina', 'QualidadeCerca', 'Outros'], axis= 1, inplace= True)

In [ ]:
# tratando a variável 'Fachada'. Vamos tirar uma média de acordo com a 'ClasseZona' para preencher os MV.

df['ClasseZona'].value_counts()

In [ ]:
df[df['ClasseZona'] == 'RL']['Fachada'].mean()


In [ ]:
df[df['ClasseZona'] == 'RM']['Fachada'].mean()

In [ ]:
df[df['ClasseZona'] == 'FV']['Fachada'].mean()

In [ ]:
df[df['ClasseZona'] == 'RH']['Fachada'].mean()

In [ ]:
df[df['ClasseZona'] == 'C (all)']['Fachada'].mean()

In [6]:
# criando os filtros para preencher os MV.

f_RL = df[df['ClasseZona'] == 'RL']['Fachada'].mean()
df['Fachada'][df['ClasseZona'] == 'RL'] = df['Fachada'].fillna(f_RL)


C:\Users\lucas\AppData\Local\Temp/ipykernel_9264/1276256780.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Fachada'][df['ClasseZona'] == 'RL'] = df['Fachada'].fillna(f_RL)


In [7]:
f_RM = df[df['ClasseZona'] == 'RM']['Fachada'].mean()
df['Fachada'][df['ClasseZona'] == 'RM'] = df['Fachada'].fillna(f_RM)

C:\Users\lucas\AppData\Local\Temp/ipykernel_9264/311017561.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Fachada'][df['ClasseZona'] == 'RM'] = df['Fachada'].fillna(f_RM)


In [8]:
f_FV = df[df['ClasseZona'] == 'FV']['Fachada'].mean()
df['Fachada'][df['ClasseZona'] == 'FV'] = df['Fachada'].fillna(f_FV)

C:\Users\lucas\AppData\Local\Temp/ipykernel_9264/1797435397.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Fachada'][df['ClasseZona'] == 'FV'] = df['Fachada'].fillna(f_FV)


In [9]:
f_RH = df[df['ClasseZona'] == 'RH']['Fachada'].mean()
df['Fachada'][df['ClasseZona'] == 'RH'] = df['Fachada'].fillna(f_RH)

C:\Users\lucas\AppData\Local\Temp/ipykernel_9264/2857994189.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Fachada'][df['ClasseZona'] == 'RH'] = df['Fachada'].fillna(f_RH)


In [10]:
f_C = df[df['ClasseZona'] == 'C (all)']['Fachada'].mean()
df['Fachada'][df['ClasseZona'] == 'C (all)'] = df['Fachada'].fillna(f_C)

C:\Users\lucas\AppData\Local\Temp/ipykernel_9264/3655048961.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Fachada'][df['ClasseZona'] == 'C (all)'] = df['Fachada'].fillna(f_C)


In [11]:
df.isnull().mean().round(4)*100

Id                      0.00
ClasseImovel            0.00
ClasseZona              0.00
Fachada                 0.00
TamanhoLote             0.00
Rua                     0.00
FormaProp               0.00
PlanoProp               0.00
Servicos                0.00
ConfigLote              0.00
InclinacaoLote          0.00
Bairro                  0.00
Estrada1                0.00
Estrada2                0.00
TipoHabitacao           0.00
EstiloHabitacao         0.00
Qualidade               0.00
Condicao                0.00
AnoConstrucao           0.00
AnoReforma              0.00
TipoTelhado             0.00
MaterialTelhado         0.00
Exterior1               0.00
Exterior2               0.00
TipoAlvenaria           0.55
AreaAlvenaria           0.55
QualidadeCobertura      0.00
CondicaoExterna         0.00
TipoFundacao            0.00
AlturaPorao             2.53
CondicaoPorao           2.53
ParedePorao             2.60
TipoAcabPorao1          2.53
AreaAcabPorao1          0.00
TipoAcabPorao2

In [12]:
# com as outras variáveis apresentam um n. pequeno de MV, vamos excluir as observações onde contem MV.
df.dropna(inplace= True)

In [ ]:
df.info()

In [ ]:
# tratamos inicialmente os dados, deixando todas as variáveis com o mesmo n. de observação.


# Análise Exploratória seguindo o modelo da TERA

In [ ]:
df['PrecoVenda'].describe()

In [ ]:
# observando o preço dos imóveis de acordo com algumas variáveis qualitativas

sns.boxplot(x= df['ClasseZona'], y= df['PrecoVenda'] )

### Maiores preços (outliers ?)estão na 'RL - Residencia baixa densidade'

In [ ]:
sns.boxplot(x= df['Qualidade'], y= df['PrecoVenda'] )

### preços de acordo com a qualidade de acabamento do imóvel

In [ ]:
sns.boxplot(y= df['ClasseImovel'], x= df['PrecoVenda'] )
# imóveis com inclinação do lote severa possuem uma mediana um pouco maior nos preços.

In [ ]:
sns.boxplot(y= df['Bairro'], x= df['PrecoVenda'] )

In [ ]:
# determinando as variáveis categóricas e numéricas
df.columns

In [ ]:
# variáveis categóricas:
['ClasseImovel', 'ClasseZona', 'Rua','FormaProp', 'PlanoProp', 'Servicos', 'ConfigLote', 'InclinacaoLote',
 'Bairro', 'Estrada1', 'Estrada2', 'TipoHabitacao', 'EstiloHabitacao', 'Qualidade', 'Condicao', 'TipoTelhado', 'MaterialTelhado',
 'Exterior1', 'Exterior2', 'TipoAlvenaria', 'QualidadeCobertura', 'CondicaoExterna', 'TipoFundacao', 'AlturaPorao',
 'CondicaoPorao', 'ParedePorao', 'TipoAcabPorao1', 'TipoAcabPorao2', 'Aquecimento', 'QualidadeAquecimento',
 'ArCentral', 'InstalacaoEletrica', 'QualidadeCozinha', 'Funcionalidade', 'LocalGaragem', 'AcabamentoGaragem',
 'QualidadeGaragem', 'CondicaoGaragem', 'EntradaPavimentada', 'TipoVenda', 'CondicaoVenda']

In [ ]:
# variáveis numéricas:
['Id', 'Fachada','TamanhoLote', 'AnoConstrucao', 'AnoReforma', 'AreaAlvenaria', 'AreaAcabPorao1', 'AreaAcabPorao2',
 'AreaInacabPorao', 'AreaPorao', 'AreaTerreo', 'Area2Andar', 'BaixaQualiAreaAcab', 'AreaConstruida', 'BanheiroPorao',
 'LavaboPorao', 'Banheiro', 'Lavabo', 'BedroomAbvGr', 'KitchenAbvGr', 'TotalQuartos', 'Lareira', 'AnoGaragem', 'CarrosGaragem',
 'AreaGaragem', 'AreaDeck', 'AreaVarandaAberta', 'AreaVarandaFechada', 'AreaVaranda3Estacoes', 'AreaAlpendre',
 'AreaPiscina', 'ValorOutros', 'MesVenda', 'AnoVenda', 'PrecoVenda']
 

In [13]:
# transformando as variáveis em categóricas
df[['ClasseImovel', 'ClasseZona', 'Rua','FormaProp', 'PlanoProp', 'Servicos', 'ConfigLote', 'InclinacaoLote',
 'Bairro', 'Estrada1', 'Estrada2', 'TipoHabitacao', 'EstiloHabitacao', 'Qualidade', 'Condicao', 'TipoTelhado', 'MaterialTelhado',
 'Exterior1', 'Exterior2', 'TipoAlvenaria', 'QualidadeCobertura', 'CondicaoExterna', 'TipoFundacao', 'AlturaPorao',
 'CondicaoPorao', 'ParedePorao', 'TipoAcabPorao1', 'TipoAcabPorao2', 'Aquecimento', 'QualidadeAquecimento',
 'ArCentral', 'InstalacaoEletrica', 'QualidadeCozinha', 'Funcionalidade', 'LocalGaragem', 'AcabamentoGaragem',
 'QualidadeGaragem', 'CondicaoGaragem', 'EntradaPavimentada', 'TipoVenda', 'CondicaoVenda']] = df[['ClasseImovel', 'ClasseZona', 'Rua','FormaProp', 'PlanoProp', 'Servicos', 'ConfigLote', 'InclinacaoLote',
 'Bairro', 'Estrada1', 'Estrada2', 'TipoHabitacao', 'EstiloHabitacao', 'Qualidade', 'Condicao', 'TipoTelhado', 'MaterialTelhado',
 'Exterior1', 'Exterior2', 'TipoAlvenaria', 'QualidadeCobertura', 'CondicaoExterna', 'TipoFundacao', 'AlturaPorao',
 'CondicaoPorao', 'ParedePorao', 'TipoAcabPorao1', 'TipoAcabPorao2', 'Aquecimento', 'QualidadeAquecimento',
 'ArCentral', 'InstalacaoEletrica', 'QualidadeCozinha', 'Funcionalidade', 'LocalGaragem', 'AcabamentoGaragem',
 'QualidadeGaragem', 'CondicaoGaragem', 'EntradaPavimentada', 'TipoVenda', 'CondicaoVenda']].astype('category')

In [14]:
# criando uma feature de areaConstruida/TamanhoLote

df['AreaConst/tamLote'] = df['AreaConstruida']/df['TamanhoLote']

In [ ]:
df.info()

In [ ]:
# Heatmap das variáveis contínuas
sns.heatmap(df[['Fachada','TamanhoLote', 'AreaAlvenaria', 'AreaAcabPorao1', 'AreaAcabPorao2',
 'AreaInacabPorao', 'AreaPorao', 'AreaTerreo', 'Area2Andar', 'BaixaQualiAreaAcab', 'AreaConstruida',
 'AreaGaragem', 'AreaDeck', 'AreaVarandaAberta', 'AreaVarandaFechada', 'AreaVaranda3Estacoes', 'AreaAlpendre',
 'AreaPiscina','AreaConst/tamLote', 'PrecoVenda']].corr().round(2), annot=True, vmin=-1, vmax=1)
sns.set(rc={"figure.figsize":(11, 9)})

### As maiores correlações com nosso alvo (PreçoVenda) são AreaPorao, AreaTerreo, AreaConstruida, Areagaragem.


In [ ]:
# heatmap das variáveis Discretas
sns.heatmap(df[['BanheiroPorao', 'LavaboPorao', 'Banheiro', 'Lavabo', 'BedroomAbvGr', 'KitchenAbvGr', 'TotalQuartos', 'Lareira', 'CarrosGaragem',
 'PrecoVenda']].corr().round(2), annot=True, vmin=-1, vmax=1)
sns.set(rc={"figure.figsize":(11, 9)})

### Maiores correlações, Banheiro, TotalQuartos, CarrosGaragem

# Objetivo 1:

In [78]:
from sklearn.model_selection import train_test_split
import statsmodels.formula.api as smf
from sklearn.preprocessing import power_transform

In [15]:
df['CondicaoVenda'].value_counts()

### novo dataset deve ter 1104 observações

Normal     1104
Partial     120
Abnorml      86
Family       20
Alloca        7
AdjLand       1
Name: CondicaoVenda, dtype: int64

In [16]:
# salvando novo dataset com o filtro 'condicaoVenda' == normal

novo_df = df[df['CondicaoVenda'] == 'Normal']

In [ ]:
novo_df.info()

In [17]:
# as variáveis selecionadas para o modelo serão: 
# AreaConst/tamLote, AreaConstruida, TotalQuartos, CarroGaragem, ClasseImóvel e Bairro  

novo_df_modelo = novo_df[['AreaConst/tamLote', 'AreaConstruida', 'TotalQuartos', 'CarrosGaragem', 'ClasseImovel', 'Bairro', 'PrecoVenda']]

In [18]:
novo_df_modelo.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1104 entries, 0 to 1459
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   AreaConst/tamLote  1104 non-null   float64 
 1   AreaConstruida     1104 non-null   int64   
 2   TotalQuartos       1104 non-null   int64   
 3   CarrosGaragem      1104 non-null   int64   
 4   ClasseImovel       1104 non-null   category
 5   Bairro             1104 non-null   category
 6   PrecoVenda         1104 non-null   int64   
dtypes: category(2), float64(1), int64(4)
memory usage: 55.3 KB


In [48]:
#dummizando

df_dummie = pd.get_dummies(novo_df_modelo, columns= ['ClasseImovel', 'Bairro'], drop_first= True)

In [51]:
df_dummie.columns

Index(['AreaConst/tamLote', 'AreaConstruida', 'TotalQuartos', 'CarrosGaragem',
       'PrecoVenda', 'ClasseImovel_30', 'ClasseImovel_40', 'ClasseImovel_45',
       'ClasseImovel_50', 'ClasseImovel_60', 'ClasseImovel_70',
       'ClasseImovel_75', 'ClasseImovel_80', 'ClasseImovel_85',
       'ClasseImovel_90', 'ClasseImovel_120', 'ClasseImovel_160',
       'ClasseImovel_180', 'ClasseImovel_190', 'Bairro_Blueste',
       'Bairro_BrDale', 'Bairro_BrkSide', 'Bairro_ClearCr', 'Bairro_CollgCr',
       'Bairro_Crawfor', 'Bairro_Edwards', 'Bairro_Gilbert', 'Bairro_IDOTRR',
       'Bairro_MeadowV', 'Bairro_Mitchel', 'Bairro_NAmes', 'Bairro_NPkVill',
       'Bairro_NWAmes', 'Bairro_NoRidge', 'Bairro_NridgHt', 'Bairro_OldTown',
       'Bairro_SWISU', 'Bairro_Sawyer', 'Bairro_SawyerW', 'Bairro_Somerst',
       'Bairro_StoneBr', 'Bairro_Timber', 'Bairro_Veenker'],
      dtype='object')

In [60]:
# separando as variáveis X
var_X = df_dummie.drop(['PrecoVenda'], axis=1)
var_X

AreaConst/tamLote  AreaConstruida  TotalQuartos  CarrosGaragem  \
0              0.202367            1710             8              2   
1              0.131458            1262             6              2   
2              0.158756            1786             6              2   
4              0.154137            2198             9              3   
5              0.096493            1362             5              2   
6              0.167989            1694             7              2   
7              0.201310            2090             7              2   
9              0.145148            1077             5              1   
10             0.092857            1040             5              1   
12             0.070327             912             4              1   
14             0.114744            1253             5              1   
15             0.139542             854             5              2   
16             0.089316            1004             5              2   
18             0.081344            1114             6              2   
21             0.148745            1108             6              1   
22             0.184254            1795             7              2   
23             0.250947            1060             6              2   
24             0.128547            1060             6              1   
25             0.112439            1600             7              3   
26             0.125000             900             5              2   
27             0.148458            1704             7              3   
28             0.098033            1600             6              1   
29             0.082226             520             4              1   
30             0.154941            1317             6              1   
31             0.143727            1228             6              1   
32             0.111684            1234             7              2   
33             0.161107            1700             6              2   
34             0.213455            1561             6              2   
35             0.182740            2452             9              3   
36             0.101022            1097             6              2   
37             0.152016            1297             5              2   
41             0.078557            1328             5              1   
42             0.096296             884             5              2   
43             0.101957             938             5              1   
44             0.144745            1150             6              1   
45             0.228780            1752             6              2   
47             0.149243            1656             7              3   
49             0.123353             955             6              1   
50             0.105992            1470             6              2   
51             0.188462            1176             6              2   
52             0.096317             816             5              2   
53             0.036641            1842             5              3   
54             0.190636            1360             6              2   
55             0.140049            1425             7              2   
57             0.147703            1720             7              2   
59             0.108333             780             4              1   
61             0.154306            1111             6              1   
62             0.212667            1370             6              2   
63             0.166019            1710             9              2   
64             0.216960            2034             8              2   
65             0.257846            2473             9              3   
66             0.110905            2207             7              2   
67             0.138678            1479             7              2   
68             0.162109             747             4              1   
69             0.146668            2287       

In [80]:
# variável Y

var_Y = df_dummie['PrecoVenda']
var_Y_log = np.log(var_Y)


In [75]:
X_train, X_test, y_train, y_test = train_test_split(var_X, var_Y_log, test_size=0.2)

In [76]:
# rodando a OLS

modelo_1 = smf.ols(formula= 'y_train ~ X_train', data= df_dummie.loc[X_train.index , :]).fit()

In [77]:
modelo_1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                y_train   R-squared:                       0.839
Model:                            OLS   Adj. R-squared:                  0.831
Method:                 Least Squares   F-statistic:                     104.2
Date:                Tue, 01 Feb 2022   Prob (F-statistic):          5.45e-301
Time:                        20:09:15   Log-Likelihood:                 491.10
No. Observations:                 883   AIC:                            -896.2
Df Residuals:                     840   BIC:                            -690.5
Df Model:                          42                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      11.5356      0.067    171.372      0.000      11.404      11.668
X_train[0]     -0.4152      0.094     -4.424      0.000      -0.599      -0.231
X_train[1]      0.0005   2.25e-05     22.727      0.000       0.000       0.001
X_train[2]     -0.0221      0.006     -3.510      0.000      -0.034      -0.010
X_train[3]      0.0908      0.011      8.344      0.000       0.069       0.112
X_train[4]     -0.2293      0.032     -7.112      0.000      -0.293      -0.166
X_train[5]     -0.1141      0.074     -1.548      0.122      -0.259       0.031
X_train[6]     -0.1982      0.051     -3.873      0.000      -0.299      -0.098
X_train[7]     -0.1754      0.023     -7.738      0.000      -0.220      -0.131
X_train[8]     -0.0795      0.017     -4.593      0.000      -0.114      -0.046
X_train[9]     -0.1903      0.031     -6.050      0.000      -0.252      -0.129
X_train[10]    -0.1793      0.053     -3.357      0.001      -0.284      -0.074
X_train[11]     0.0058      0.025      0.227      0.820      -0.044       0.055
X_train[12]     0.0341      0.039      0.883      0.377      -0.042       0.110
X_train[13]    -0.2602      0.037     -7.110      0.000      -0.332      -0.188
X_train[14]     0.0063      0.029      0.217      0.829      -0.051       0.063
X_train[15]    -0.1069      0.052     -2.059      0.040      -0.209      -0.005
X_train[16]     0.0503      0.078      0.643      0.520      -0.103       0.204
X_train[17]    -0.2540      0.044     -5.728      0.000      -0.341      -0.167
X_train[18]    -0.0117      0.117     -0.100      0.921      -0.242       0.218
X_train[19]    -0.1442      0.075     -1.926      0.054      -0.291       0.003
X_train[20]    -0.1451      0.061     -2.377      0.018      -0.265      -0.025
X_train[21]    -0.0751      0.066     -1.136      0.256      -0.205       0.055
X_train[22]    -0.0779      0.055     -1.427      0.154      -0.185       0.029
X_train[23]     0.0066      0.061      0.109      0.914      -0.112       0.126
X_train[24]    -0.2595      0.058     -4.450      0.000      -0.374      -0.145
X_train[25]    -0.1270      0.058     -2.195      0.028      -0.241      -0.013
X_train[26]    -0.2588      0.065     -3.958      0.000      -0.387      -0.130
X_train[27]    -0.3728      0.074     -5.037      0.000      -0.518      -0.228
X_train[28]    -0.1657      0.059     -2.821      0.005      -0.281      -0.050
X_train[29]    -0.2254      0.055     -4.066      0.000      -0.334      -0.117
X_train[30]    -0.1150      0.074     -1.557      0.120      -0.260       0.030
X_train[31]    -0.1846      0.058     -3.188      0.001      -0.298      -0.071
X_train[32]    -0.0487      0.061     -0.801      0.423      -0.168       0.071
X_train[33]     0.0940      0.055      1.715      0.087      -0.014       0.202
X_train[34]    -0.2854      0.058     -4.95

Nosso modelo apresento um R quadrado de 0.839. 
O teste de Durbin-Watson demonstrou que não existe autocorrelação.

